File: frontend/lib/pages/add_blocks_and_stairs.dart
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';

import '../config/environment.dart';

class AddBlocksAndStairsPage extends StatefulWidget {
  final int associationId;

  const AddBlocksAndStairsPage({Key? key, required this.associationId})
      : super(key: key);

  @override
  _AddBlocksAndStairsPageState createState() => _AddBlocksAndStairsPageState();
}

class _AddBlocksAndStairsPageState extends State<AddBlocksAndStairsPage> {
  final _blockNameController = TextEditingController();
  final _stairNameController = TextEditingController();
  final List<Map<String, dynamic>> _blocks = [];
  final List<String> _stairs = [];
  String? _selectedBlockId;
  bool _isLoading = false;
  bool _blocksSubmitted = false;

  Future<void> _loadBlocks() async {
    setState(() {
      _isLoading = true;
    });

    final prefs = await SharedPreferences.getInstance();
    final accessToken = prefs.getString('access_token');

    final response = await http.get(
      Uri.parse('${EnvironmentConfig.API_URL}/blocks?associationId=${widget.associationId}'),
      headers: {
        'Authorization': 'Bearer $accessToken',
      },
    );

    setState(() {
      _isLoading = false;
    });

    if (response.statusCode == 200) {
      final List<dynamic> responseBody = jsonDecode(response.body);
      setState(() {
        _blocks.clear();
        _blocks.addAll(responseBody.map((block) {
          return {
            'id': block['id'],
            'name': block['name'],
            'stairs': block['stairs'] ?? [],
          };
        }).toList());
      });
    } else {
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(content: Text('Eșec la încărcarea blocurilor')),
      );
    }
  }

  void _addBlock() {
    if (_blockNameController.text.isNotEmpty) {
      setState(() {
        _blocks.add({'name': _blockNameController.text, 'id': null, 'stairs': []});
        _blockNameController.clear();
      });
    }
  }

  Future<void> _submitBlocks() async {
    final prefs = await SharedPreferences.getInstance();
    final accessToken = prefs.getString('access_token');

    final response = await http.post(
      Uri.parse('${EnvironmentConfig.API_URL}/addBlocks?associationId=${widget.associationId}'),
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $accessToken',
      },
      body: jsonEncode(
          _blocks.map((block) => {'name': block['name']}).toList()),
    );

    if (response.statusCode == 201) {
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(content: Text('Blocuri adăugate cu succes')),
      );
      setState(() {
        _blocksSubmitted = true;
      });

      await _loadBlocks();
    } else {
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(content: Text('Eșec la adăugarea blocurilor')),
      );
    }
  }

  Future<void> _addStair() async {
    if (_selectedBlockId == null) {
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(content: Text('Te rog selectează un bloc')),
      );
      return;
    }

    final prefs = await SharedPreferences.getInstance();
    final accessToken = prefs.getString('access_token');

    final response = await http.post(
      Uri.parse('${EnvironmentConfig.API_URL}/addStair?blockId=$_selectedBlockId'),
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $accessToken',
      },
      body: jsonEncode(_stairs.map((name) => {'name': name}).toList()),
    );

    if (response.statusCode == 201) {
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(content: Text('Scările adăugate cu succes')),
      );
      setState(() {
        _stairs.clear();
      });

      await _loadBlocks();
    } else {
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(content: Text('Eșec la adăugarea scărilor')),
      );
    }
  }

  Widget buildTextField(String labelText, TextEditingController controller,
      {double? width}) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text(labelText, style: TextStyle(fontSize: 14, color: Colors.black)),
        SizedBox(height: 4),
        Container(
          width: width ?? 200,
          decoration: BoxDecoration(
            color: Color(0xFFECF4FF),
            borderRadius: BorderRadius.circular(5),
          ),
          child: TextFormField(
            controller: controller,
            decoration: InputDecoration(
              filled: true,
              fillColor: Color(0xFFECF4FF),
              border: OutlineInputBorder(
                borderSide: BorderSide(color: Colors.transparent),
                borderRadius: BorderRadius.all(Radius.circular(5)),
              ),
              enabledBorder: OutlineInputBorder(
                borderSide: BorderSide(color: Colors.transparent),
                borderRadius: BorderRadius.all(Radius.circular(5)),
              ),
              focusedBorder: OutlineInputBorder(
                borderSide: BorderSide(color: Colors.blue),
                borderRadius: BorderRadius.all(Radius.circular(5)),
              ),
            ),
            style: TextStyle(fontSize: 14),
          ),
        ),
      ],
    );
  }

  @override
  void initState() {
    super.initState();
    _loadBlocks();
  }

  @override
  Widget build(BuildContext context) {
    Size size = MediaQuery.of(context).size;

    return Scaffold(
      backgroundColor: Colors.white,
      body: SafeArea(
        child: SingleChildScrollView(
          child: Container(
            padding: EdgeInsets.symmetric(
                horizontal: size.width > 600 ? 150 : 70, vertical: 50),
            child: Row(
              children: [
                Expanded(
                  flex: 3,
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      SizedBox(height: size.height * 0.02),
                      Text(
                        'Adaugă Blocuri și Scări',
                        style: TextStyle(
                          fontSize: 24,
                          fontWeight: FontWeight.bold,
                          color: Colors.purple,
                        ),
                      ),
                      SizedBox(height: 60),
                      Text(
                        '1. Adaugă Blocuri',
                        style: TextStyle(
                          fontSize: 18,
                          fontWeight: FontWeight.bold,
                        ),
                      ),
                      SizedBox(height: 16),
                      Row(
                        mainAxisAlignment: MainAxisAlignment.start,
                        children: [
                          Container(
                            width: 200,
                            child: buildTextField('Nume Bloc', _blockNameController),
                          ),
                          Padding(
                            padding: EdgeInsets.only(top: 25),
                            child: IconButton(
                              icon: Icon(Icons.add),
                              onPressed: _addBlock,
                            ),
                          ),
                        ],
                      ),
                      SizedBox(height: 10),
                      for (var block in _blocks)
                        Padding(
                          padding: const EdgeInsets.symmetric(vertical: 4.0),
                          child: Container(
                            width: 150,
                            decoration: BoxDecoration(
                              border: Border.all(color: Colors.grey),
                              borderRadius: BorderRadius.circular(5),
                            ),
                            padding: EdgeInsets.all(8),
                            child: Text(block['name']),
                          ),
                        ),
                      SizedBox(height: 16),
                      ElevatedButton(
                        onPressed: _submitBlocks,
                        child: Text('Trimite Blocurile'),
                      ),
                      SizedBox(height: 60),

                      if (_blocksSubmitted) ...[
                        Text(
                          '2. Adaugă Scările',
                          style: TextStyle(
                            fontSize: 18,
                            fontWeight: FontWeight.bold,
                          ),
                        ),
                        SizedBox(height: 16),
                        Container(
                          width: 200,
                          decoration: BoxDecoration(
                            color: Colors.white,
                            borderRadius: BorderRadius.circular(5),
                          ),
                          child: DropdownButtonFormField<String>(
                            value: _selectedBlockId,
                            items: _blocks.map((block) {
                              return DropdownMenuItem<String>(
                                value: block['id'].toString(),
                                child: Text(block['name']),
                              );
                            }).toList(),
                            onChanged: (value) {
                              setState(() {
                                _selectedBlockId = value;
                              });
                            },
                            decoration: InputDecoration(
                              labelText: 'Alege un bloc',
                              filled: true,
                              fillColor: Color(0xFFECF4FF),
                              border: OutlineInputBorder(
                                borderSide: BorderSide(color: Colors.transparent),
                              ),
                            ),
                          ),
                        ),
                        SizedBox(height: 16),
                        Row(
                          children: [
                            Container(
                              width: 200,
                              child: buildTextField('Nume Scară', _stairNameController),
                            ),
                            Padding(
                              padding: EdgeInsets.only(top: 25),
                              child: IconButton(
                                icon: Icon(Icons.add),
                                onPressed: () {
                                  setState(() {
                                    _stairs.add(_stairNameController.text);
                                    _stairNameController.clear();
                                  });
                                },
                              ),
                            ),
                          ],
                        ),
                        SizedBox(height: 10),
                        for (var stair in _stairs)
                          Padding(
                            padding: const EdgeInsets.symmetric(vertical: 4.0),
                            child: Container(
                              width: 150,
                              decoration: BoxDecoration(
                                border: Border.all(color: Colors.grey),
                                borderRadius: BorderRadius.circular(5),
                              ),
                              padding: EdgeInsets.all(8),
                              child: Text(stair),
                            ),
                          ),
                        SizedBox(height: 16),
                        ElevatedButton(
                          onPressed: _addStair,
                          child: Text('Adaugă Scările'),
                        ),
                      ],
                    ],
                  ),
                ),
                SizedBox(width: 20),
                Expanded(
                  flex: 5,
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(
                        'Lista Blocuri și Scări',
                        style: TextStyle(
                          fontSize: 20,
                          fontWeight: FontWeight.bold,
                          color: Colors.purple,
                        ),
                      ),
                      SizedBox(height: 16),
                      Container(
                        height: 700,
                        child: ListView.builder(
                          itemCount: _blocks.length,
                          itemBuilder: (context, index) {
                            final block = _blocks[index];
                            return Column(
                              crossAxisAlignment: CrossAxisAlignment.start,
                              children: [
                                Text(
                                  'Bloc: ${block['name']}',
                                  style: TextStyle(
                                      fontSize: 18, fontWeight: FontWeight.bold),
                                ),
                                for (var stair in block['stairs'])
                                  Padding(
                                    padding: const EdgeInsets.only(left: 16, top: 4),
                                    child: Text(
                                      'Scara: ${stair['name']}',
                                      style: TextStyle(
                                          fontSize: 16, color: Colors.black54),
                                    ),
                                  ),
                                SizedBox(height: 12),
                              ],
                            );
                          },
                        ),
                      ),
                    ],
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}

File: web/src/pages/features/association/add-blocks-and-stairs.js
import React, { useState, useEffect, useCallback } from "react";
import axios from "axios";
import { useParams } from "react-router-dom";
import Sidebar from "../components/sidemenu";
import "../../../pages/features/styles/addblocksandstairs.css";

const AddBlocksAndStairsPage = () => {
  const { associationId } = useParams();
  const [blockName, setBlockName] = useState("");
  const [stairName, setStairName] = useState("");
  const [blocks, setBlocks] = useState([]);
  const [selectedBlockId, setSelectedBlockId] = useState(null);
  const [isLoading, setIsLoading] = useState(false);
  const [blocksSubmitted, setBlocksSubmitted] = useState(false);

  const loadBlocks = useCallback(async () => {
    setIsLoading(true);
    const token = localStorage.getItem("access_token");
    try {
      const apiUrl = process.env.REACT_APP_API_URL;
      const response = await axios.get(
        `${apiUrl}/blocks?associationId=${associationId}`,
        {
          headers: {
            Authorization: `Bearer ${token}`,
          },
        }
      );
      setBlocks(response.data.map(block => ({
        id: block.id,
        name: block.name,
        stairs: block.stairs || [],
      })));
    } catch (error) {
      console.error("Eroare la încărcarea blocurilor:", error);
    } finally {
      setIsLoading(false);
    }
  }, [associationId]);

  useEffect(() => {
    loadBlocks();
  }, [associationId, loadBlocks]);

  const addBlock = async () => {
    if (blockName.trim() === "") return;
    const token = localStorage.getItem("access_token");
    try {
      const apiUrl = process.env.REACT_APP_API_URL;
      await axios.post(
        `${apiUrl}/addBlocks?associationId=${associationId}`,
        [{ name: blockName }],
        {
          headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${token}`,
          },
        }
      );
      setBlockName("");
      setBlocksSubmitted(true);  
      loadBlocks();

     
      setTimeout(() => {
        setBlocksSubmitted(false);
      }, 3000);
    } catch (error) {
      console.error("Eroare la adăugarea blocului:", error);
    }
  };

  const addStair = async () => {
    if (!selectedBlockId || stairName.trim() === "") return;
    const token = localStorage.getItem("access_token");
    try {
      const apiUrl = process.env.REACT_APP_API_URL;
      await axios.post(
        `${apiUrl}/addStair?blockId=${selectedBlockId}`,
        [{ name: stairName }],
        {
          headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${token}`,
          },
        }
      );
      setStairName("");
      loadBlocks();
    } catch (error) {
      console.error("Eroare la adăugarea scării:", error);
    }
  };

  return (
    <div className="add-blocks-container-unique-all">
      <div className="add-blocks-container-unique">
        <Sidebar />
        <div className="add-blocks-left-unique">
          <h2 className="add-blocks-title-unique">1. Adaugă Blocuri</h2>
          <div className="form-row-unique">
            <input
              type="text"
              placeholder="Nume Bloc"
              value={blockName}
              onChange={(e) => setBlockName(e.target.value)}
              className="block-name-input-unique"
            />
            <button onClick={addBlock} className="add-block-button-unique">Adaugă Bloc</button>
          </div>
          {isLoading && <p>Se încarcă blocurile...</p>}
          {blocksSubmitted && <p>Blocul a fost adăugat cu succes!</p>}
        </div>

        <div className="add-blocks-right-unique">
          <h2 className="add-stairs-title-unique">2. Adaugă Scări</h2>
          <select
            value={selectedBlockId}
            onChange={(e) => setSelectedBlockId(e.target.value)}
            className="block-select-unique"
          >
            <option value="">Selectează un bloc</option>
            {blocks.map((block) => (
              <option key={block.id} value={block.id}>
                {block.name}
              </option>
            ))}
          </select>
          <div className="form-row-unique">
            <input
              type="text"
              placeholder="Nume Scară"
              value={stairName}
              onChange={(e) => setStairName(e.target.value)}
              className="stair-name-input-unique"
            />
            <button onClick={addStair} className="add-stair-button-unique">Adaugă Scară</button>
          </div>
        </div>
      </div>

      <div className="add-blocks-list-container">
        <h3>Blocuri și Scări Adăugate:</h3>
        <div className="add-blocks-list">
          {blocks.length > 0 ? (
            blocks.map((block) => (
              <div key={block.id} className="block-details">
                <p><strong>Bloc:</strong> {block.name}</p>
                <p><strong>Scări:</strong> {block.stairs.map((stair, index) => (
                  <span key={index}>{stair.name} </span>
                ))}</p>
              </div>
            ))
          ) : (
            <p>Nu există blocuri sau scări adăugate.</p>
          )}
        </div>
      </div>
    </div>
  );
};

export default AddBlocksAndStairsPage;
